token-metadata/program/src/processor/collection/unverify_collection.rs

```diff
@@ -1,71 +1,115 @@
+    assert_owned_by(collection_mint_info, &spl_token::id())?;
 
+    // Deserialize the collection item metadata.
     let mut metadata = Metadata::from_account_info(metadata_info)?;
-    let collection_data = Metadata::from_account_info(collection_info)?;
-
-    assert_collection_verify_is_valid(
-        &metadata.collection,
-        &collection_data,
-        collection_mint,
-        edition_account_info,
-    )?;
-    if using_delegated_collection_authority {
-        let collection_authority_record = next_account_info(account_info_iter)?;
-        assert_has_collection_authority(
-            collection_authority_info,
-            &collection_data,
-            collection_mint.key,
-            Some(collection_authority_record),
-        )?;
-    } else {
-        assert_has_collection_authority(
-            collection_authority_info,
-            &collection_data,
-            collection_mint.key,
-            None,
-        )?;
+
+    // First, if there's no collection set, we can just short-circuit
+    // since there's nothing to unverify.
+    let collection = match metadata.collection.as_mut() {
+        Some(collection) => collection,
+        None => return Ok(()),
+    };
+
+    // Short-circuit if it's already unverified.
+    if !collection.verified {
+        return Ok(());
     }
 
-    // This handler can only unverify non-sized NFTs
-    if collection_data.collection_details.is_some() {
-        return Err(MetadataError::SizedCollection.into());
+    // The collection parent must be the actual parent of the
+    // collection item.
+    if collection.key != *collection_mint_info.key {
+        return Err(MetadataError::NotAMemberOfCollection.into());
     }
 
-    // If the NFT has collection data, we set it to be unverified and then update the collection
-    // size on the Collection Parent.
-    if let Some(collection) = &mut metadata.collection {
-        collection.verified = false;
+    // We need to ensure the metadata is derived from the mint so
+    // someone cannot pass in a burned metadata account associated with
+    // a different mint.
+    assert_metadata_derivation(program_id, collection_metadata_info, collection_mint_info)?;
+
+    // Check if the collection metadata account is burned. If it is,
+    // there's no sized data to update and the user can simply unverify
+    // the NFT.
+    //
+    // This check needs to happen before the program owned check.
+    let parent_burned = collection_metadata_info.data_is_empty()
+        && collection_metadata_info.owner == &system_program::ID;
+
+    if parent_burned {
+        // If the parent is burned, we need to check that the authority
+        // is the update authority on the item metadata.
+        //
+        // Collection Delegates for burned collection parents should not be
+        // respected as there's currently no way to revoke them.
+
+        if metadata.update_authority != *collection_authority_info.key {
```

Comment:
Changing Shank annotation names is a breaking API change, though, which I'd like to avoid. I can change the name internally to be more clear but don't want to change the name in the instruction API in either Rust or JS.

Reply:
If its internal then maybe its the scenario where you are looking at the code anyways, so maybe it doesn't add much of a benefit.  I understand about not breaking API change with shank annotation